test/distributed/_composable/fsdp/test_fully_shard_memory.py

```diff
@@ -0,0 +1,157 @@
+# Owner(s): ["oncall: distributed"]
+
+import functools
+
+import torch
+
+from torch.distributed._composable.fsdp import fully_shard
+from torch.testing._internal.common_distributed import skip_if_lt_x_gpu
+from torch.testing._internal.common_fsdp import FSDPTest
+from torch.testing._internal.common_utils import run_tests
+from torch.testing._internal.distributed._tensor.common_dtensor import (
+    ModelArgs,
+    Transformer,
+    TransformerBlock,
+)
+
+
+class TestFullyShardMemory(FSDPTest):
+    @property
+    def world_size(self) -> int:
+        return min(2, torch.cuda.device_count())
+
+    @skip_if_lt_x_gpu(2)
+    def test_fully_shard_training_memory(self):
+        self.run_subtests(
+            {"reshard_after_forward": [True, False]},
+            self._test_fully_shard_training_memory,
+        )
+
+    def _test_fully_shard_training_memory(self, reshard_after_forward: bool):
+        assert (
+            self.world_size == 2
+        ), f"Requires world size of 2 since some values are hard coded: {self.world_size}"
+        torch.manual_seed(42)
+        # Pre-run a linear forward (gemm and bias) and backward (gemm) to
```

Comment:
@weifengpy I did this trick here 😄 

I pre-allocate the cuBLAS workspaces and do not count them to the memory by recording the baseline memory after the workspaces have been initialized.

Reply:
nice! the memory can be fully estimated now. amazing!